mopro-ffi/src/ashlang/mod.rs

```diff
@@ -0,0 +1,70 @@
+use std::fs;
+
+use ashlang::AshlangProver;
+use ashlang::SpartanProver;
+
+use super::GenerateProofResult;
+
+#[macro_export]
+macro_rules! ashlang_spartan_app {
+    () => {
+        fn generate_ashlang_spartan_proof(
+            ar1cs_path: String, // path to ar1cs file
+            secret_inputs: Vec<String>,
+        ) -> Result<mopro_ffi::GenerateProofResult, mopro_ffi::MoproError> {
+            mopro_ffi::ashlang::prove(&ar1cs_path, secret_inputs).map_err(|e| {
+                mopro_ffi::MoproError::AshlangError(
+                    "error generating ashlang spartan proof".to_string(),
+                )
+            })
+        }
+
+        fn verify_ashlang_spartan_proof(
+            ar1cs_path: String,
+            proof: Vec<u8>,
+        ) -> Result<bool, mopro_ffi::MoproError> {
+            mopro_ffi::ashlang::verify(&ar1cs_path, proof).map_err(|e| {
+                mopro_ffi::MoproError::AshlangError("error verifying proof".to_string())
+            })
+        }
+    };
+}
+
+/// Generates a spartan proof from an ar1cs file compiled
+/// with the ashlang compiler.
+pub fn prove(
+    ar1cs_path: &str, // path to ar1cs file
+    secret_inputs: Vec<String>,
+) -> anyhow::Result<GenerateProofResult> {
+    let ir_source = fs::read_to_string(ar1cs_path)?;
+    // we pass an empty vec for public inputs because
+    // they are not supported in the ashlang spartan prover
+    // outputs are public and should be used instead
+    let proof = SpartanProver::prove_ir(&ir_source, vec![], secret_inputs.clone())?;
+
+    Ok(GenerateProofResult {
+        proof: bincode::serialize(&proof)?,
+        inputs: bincode::serialize(&secret_inputs)?,
```

Comment:
It is not used in the `verify()`?

Reply:
Yes, this is the shared `GenerateProofResult` object.

Reply:
so is there a concept of public signal?

Reply:
Ah, these `inputs` are meant to be public? There is no concept of public signal in Spartan i think, but ashlang does have support for such a thing (TritonVM for example has public inputs and secret inputs).

Reply:
It seems mentioned the `public inputs` in the readme
https://github.com/microsoft/Spartan?tab=readme-ov-file
but you can do it in the future PR